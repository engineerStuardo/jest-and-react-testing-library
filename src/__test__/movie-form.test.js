import React from 'react';
import { render, screen, fireEvent, wait } from '@testing-library/react';

import MovieForm from '../components/movie-form';

global.fetch = require('jest-fetch-mock');

const emptyMovie = {
  title: '',
  description: '',
};

const movie = {
  id: 1,
  title: 'Batman the dark knight',
  description: 'Testing description in mock data',
};

describe('Movie form component', () => {
  test('Should match snapshot', () => {
    const { container } = render(<MovieForm movie={emptyMovie} />);
    expect(container).toMatchSnapshot();
  });

  test('Should have form elements', () => {
    render(<MovieForm movie={emptyMovie} />);

    const title = screen.getByLabelText(/title/i);
    expect(title).toBeTruthy();

    const description = screen.getByLabelText(/description/i);
    expect(description).toBeTruthy();

    const button = screen.getByRole('button', { name: /create/i });
    expect(button).toBeTruthy();
  });

  test('display form elements with movie data', () => {
    const { debug } = render(<MovieForm movie={movie} />);
    // debug(screen.getByLabelText('Title'));

    const titleInput = screen.getByLabelText(/title/i).value;
    expect(titleInput).toBe(movie.title);

    const descriptionInput = screen.getByLabelText(/description/i).value;
    expect(descriptionInput).toBe(movie.description);

    const button = screen.getByRole('button', { name: /update/i });
    expect(button).toBeTruthy();
  });

  test('Should trigger API request when clicked on button', async () => {
    const updatedMovie = jest.fn();

    jest
      .spyOn(global, 'fetch')
      .mockImplementationOnce(() =>
        Promise.resolve({ json: () => Promise.resolve(movie) })
      );

    render(<MovieForm movie={movie} updatedMovie={updatedMovie} />);
    const submitButton = screen.getByRole('button', { name: /update/i });
    fireEvent.click(submitButton);

    await wait(() => {
      expect(updatedMovie).toBeCalledTimes(1);
    });
  });

  test('Shouldn trigger API request when clicked on button with empty form', async () => {
    const updatedMovie = jest.fn();

    fetch.mockResponseOnce(JSON.stringify(emptyMovie));

    render(<MovieForm movie={emptyMovie} updatedMovie={updatedMovie} />);
    const submitButton = screen.getByRole('button', { name: /create/i });
    fireEvent.click(submitButton);

    await wait(() => {
      expect(updatedMovie).toBeCalledTimes(0);
    });
  });

  test('Should trigger API request when clicked on button on new movie button', async () => {
    const movieCreated = jest.fn();

    fetch.mockResponseOnce(JSON.stringify(movie));

    render(<MovieForm movie={emptyMovie} movieCreated={movieCreated} />);

    const titleInput = screen.getByLabelText(/title/i);
    const descriptionInput = screen.getByLabelText(/description/i);
    fireEvent.change(titleInput, { target: { value: 'Title1' } });
    fireEvent.change(descriptionInput, { target: { value: 'Description1' } });

    const submitButton = screen.getByRole('button', { name: /create/i });
    fireEvent.click(submitButton);

    await wait(() => {
      // expect(movieCreated.mock.calls[0][0]).toStrictEqual(movie);
      expect(movieCreated).toBeCalledWith(movie);
    });
  });
});
