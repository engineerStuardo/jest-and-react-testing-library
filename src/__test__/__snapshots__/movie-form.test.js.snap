// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie form component Should match snapshot 1`] = `
<div>
  <div
    data-testid="movie-form"
  >
    <label
      for="title"
    >
      Title
    </label>
    <br />
    <input
      id="title"
      placeholder="title"
      type="text"
      value=""
    />
    <br />
    <label
      for="description"
    >
      Description
    </label>
    <br />
    <textarea
      id="description"
      placeholder="Description"
      type="text"
    />
    <br />
    <button
      disabled=""
    >
      Create
    </button>
  </div>
</div>
`;
